src/symbol/shaping.js

```diff
@@ -479,15 +496,20 @@ function shapeLines(shaping: Shaping,
             const section = line.getSection(i);
             const sectionIndex = line.getSectionIndex(i);
             const codePoint = line.getCharCode(i);
-            // We don't know the baseline, but since we're laying out
-            // at 24 points, we can calculate how much it will move when
-            // we scale up or down.
-            const baselineOffset = (lineMaxScale - section.scale) * 24;
+
             const positions = glyphMap[section.fontStack];
             const glyph = positions && positions[codePoint];
-
             if (!glyph) continue;
 
+            // In order to make different fonts aligned, they must share a general baseline that starts from the midline
+            // of each font face.  Baseline offset is the vertical distance from font face's baseline to its top most
+            // position, which is the half size of the sum (ascender + descender). Since glyph's position is counted
+            // from the top left corner, the negative shift is needed. So different fonts shares the same baseline but
+            // with different offset shift. If font's baseline is not applicable, fall back to use a default baseline
```

Comment:
> So different fonts shares the same baseline but with different offset shift.

Why do we want different offsets for different fonts? 

Or did you mean different characters have different offsets?

Reply:
<img width="649" alt="Screen Shot 2019-09-26 at 7 44 50 PM" src="https://user-images.githubusercontent.com/11026055/65708768-01b80a00-e098-11e9-836c-c26867b7d4c7.png">

For your question:
> Why do we want different offsets for different fonts?

Taking the above picture as example, the blue box is a shaping box.  There are two font types. `ascender` and `descender` are font-level attributes, for each font `ascender` is the distance above the baseline, and `descender` is the negative distance below the baseline. As you can see, these two fonts have different baseline and different `ascender` and `descender`. But our goal is to make the two character aligned visually, we could choose different alignment, align from the bottom line or align form the middle line. Here we choose using the middle line, it means we need to put each character around the middle line of the shaping box, they share the midline as a virtual baseline. 

When processing `shapeLines` function,  for each character, we create a `positionedGlyph` with its top-left coordinates (x, y) value in the axis plane. For value of y, which is actually the same height as `ascender`. As we can see the middle line of each glyph is (`ascender` - `descender`) / 2. (`descender` is negative). The offset of `ascender` along Y axis is (`ascender` - `descender`) / 2, since it is against the positive direction, the offset is  -(`ascender` - `descender`) / 2. 

> Or did you mean different characters have different offsets?

Since `ascender` and `descender` are font level attributes, so the offset will be different for different fonts. 

But later when we create glyph, the actual top of each character will be taken into account.

Reply:
Oh, I see! Thank you for the detailed explanation, @zmiao, this is really helpful.